kategory-effects-kotlinx/src/main/kotlin/kategory/effects/data/DeferredKW.kt

```diff
@@ -0,0 +1,96 @@
+package kategory
+
+import kategory.effects.data.internal.Platform
+import kotlinx.coroutines.experimental.*
+import kotlin.coroutines.experimental.CoroutineContext
+
+typealias DeferredResult<A> = Either<Throwable, A>
+
+@higherkind
+@deriving(Monad::class, AsyncContext::class)
+data class DeferredKW<out A>(val thunk: (CoroutineContext) -> Deferred<A>) : DeferredKWKind<A> {
+
+    fun <B> map(f: (A) -> B): DeferredKW<B> =
+            flatMap { a: A -> pure(f(a)) }
+
+    fun <B> flatMap(f: (A) -> DeferredKWKind<B>): DeferredKW<B> =
+            DeferredKW { context: CoroutineContext ->
+                this@DeferredKW.attempt(context).fold(
+                        { raiseError<B>(it).runDeferred(context) },
+                        { f(it).ev().runDeferred(context) })
+            }
+
+    companion object {
+        inline operator fun <A> invoke(noinline f: suspend CoroutineScope.() -> A): DeferredKW<A> =
+                DeferredKW { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            f()
+                        }
+                    }
+                }
+
+        inline fun <A> runAsync(noinline fa: Proc<A>): DeferredKW<A> =
+                DeferredKW { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            IO.async(fa).unsafeRunSync()
+                        }
+                    }
+                }
+
+        inline fun <A> pure(a: A): DeferredKW<A> =
+                DeferredKW.invoke { a }
+
+        inline fun <A> raiseError(t: Throwable): DeferredKW<A> =
+                DeferredKW { throw t }
+
+        fun <A, B> tailRecM(a: A, f: (A) -> DeferredKWKind<Either<A, B>>): DeferredKW<B> {
+            tailrec fun go(coroutineContext: CoroutineContext, a: A, f: (A) -> DeferredKWKind<Either<A, B>>): DeferredKW<B> {
+                val result: DeferredResult<Either<A, B>> = f(a).attempt(coroutineContext)
+                /* FIXME(paco): KT-20075 If you remove return here tailrec stops working. Jetbrains Please. */
```

Comment:
I think that makes sense because otherwise you would not know what the stop condition in the translated loop would be. what it would be good to have by default in Kotlin is to have the last statement ina  block always be the return value. Explicit returns suck.